packages/target-tauri/translationfn/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+    #[error("Plural rule not found {0}")]
+    PluralRuleNotFound(&'static str),
+    #[error("Invalid RegEx Pattern")]
+    RegEx(#[from] regex::Error),
+}
+
+impl TranslationEngine {
+    pub fn new(
+        messages: HashMap<String, HashMap<String, String>>,
+        langid: &str,
+    ) -> Result<Self, Error> {
+        let langid: LanguageIdentifier = langid
+            .split("_")
+            .next()
+            .ok_or(Error::LangIdMalformed)?
+            .parse()
+            .map_err(Error::LangIdParsingFailed)?;
+        let pr = PluralRules::create(langid, PluralRuleType::CARDINAL)
+            .map_err(Error::PluralRuleNotFound)?;
+        let var_finder = Regex::new(r"%\d\$[\w\d]")?;
+
+        Ok(TranslationEngine {
+            messages,
+            plural_rules: pr,
+            var_finder,
+        })
+    }
+    pub fn translate(&self, key: &str, substitution: Substitution) -> String {
+        if let Some(entry) = self.messages.get(key) {
+            match substitution {
+                Substitution::None => {
+                    if let Some(message) = entry.get("message") {
+                        message.to_owned()
+                    } else {
+                        error!("Message not existing for {key}");
+                        key.to_owned()
+                    }
+                }
+                Substitution::String(items) => {
+                    //TODO: remove unwrap, use log instead
+                    if let Some(message) = entry.get("message") {
+                        let mut msg = message.clone();
+                        for matches in self.var_finder.find_iter(message) {
+                            let from = matches.as_str();
+                            let position: u32 = from.chars().nth(1).unwrap().to_digit(10).unwrap();
+                            msg = msg.replace(from, items.get(position as usize - 1).unwrap());
+                        }
+                        msg
+                    } else {
+                        error!("Message not existing for {key}");
+                        key.to_owned()
+                    }
+                }
+                Substitution::Quantity(quantity) => {
+                    let quantity_key = match self.plural_rules.select(quantity) {
+                        Ok(PluralCategory::ZERO) => "zero",
+                        Ok(PluralCategory::ONE) => "one",
+                        Ok(PluralCategory::TWO) => "two",
+                        Ok(PluralCategory::FEW) => "few",
+                        Ok(PluralCategory::MANY) => "many",
+                        Ok(PluralCategory::OTHER) => "other",
+                        Err(_) => {
+                            error!("Failed to  {quantity}");
+                            return key.to_owned();
+                        }
+                    };
+                    if let Some(message) = entry.get(quantity_key) {
+                        message.replace("%d", &quantity.to_string())
```

Comment:
So this won't work for strings with multiple things to substitute, and also with some strings with a single parameter, e.g. `Export all %1$d profiles`.

TBH I'm not sure about having an implementation that neither is used nor does work properly.
Maybe mentioning it in the comments would be alright.

Reply:
afaik we don't have quantity translations with multiple fields to substitute, or do we?

Reply:
the "%d" should probably also handle the "%1$d" case, so we should adjust it, maybe:
```suggestion
                        let quantity_string = quantity.to_string();
                        message.replace("%1$d", & quantity_string).replace("%d", & quantity_string)
```